tests/test_trustme.py

```diff
@@ -48,54 +55,161 @@ def test_basics():
+        assert f.read() == test_data
+
+    # append=True appends
+    with (tmpdir / "test2").open("wb") as f:
+        f.write(b"asdf")
+    b.write_to_path(str(tmpdir / "test2"), append=True)
+    with (tmpdir / "test2").open("rb") as f:
+        assert f.read() == b"asdf" + test_data
+
+    # tempfile
+    with b.tempfile(dir=str(tmpdir)) as path:
+        assert path.startswith(str(tmpdir))
+        assert path.endswith(".pem")
+        with open(path, "rb") as f:
+            assert f.read() == test_data
+
+
+def check_connection_end_to_end(wrap_client, wrap_server):
     # Client side
-    def fake_ssl_client(raw_client_sock):
+    def fake_ssl_client(ca, raw_client_sock, hostname):
         try:
-            ssl_ctx = ca.stdlib_client_context()
-            wrapped_client_sock = ssl_ctx.wrap_socket(
-                raw_client_sock, server_hostname="my-test-host.example.org")
-            print("Client got server cert:", wrapped_client_sock.getpeercert())
-            peercert = wrapped_client_sock.getpeercert()
-            san = peercert["subjectAltName"]
-            assert san == (("DNS", "my-test-host.example.org"),)
+            wrapped_client_sock = wrap_client(ca, raw_client_sock, hostname)
             # Send and receive some data to prove the connection is good
             wrapped_client_sock.send(b"x")
             assert wrapped_client_sock.recv(1) == b"y"
         except:  # pragma: no cover
             sys.excepthook(*sys.exc_info())
             raise
+        finally:
+            raw_client_sock.close()
 
     # Server side
-    def fake_ssl_server(raw_server_sock):
+    def fake_ssl_server(server_cert, raw_server_sock):
         try:
-            ssl_ctx = server_cert.stdlib_server_context()
-            wrapped_server_sock = ssl_ctx.wrap_socket(
-                raw_server_sock, server_side=True)
+            wrapped_server_sock = wrap_server(server_cert, raw_server_sock)
             # Prove that we're connected
-            print("server encrypted with:", wrapped_server_sock.cipher())
             assert wrapped_server_sock.recv(1) == b"x"
             wrapped_server_sock.send(b"y")
         except:  # pragma: no cover
             sys.excepthook(*sys.exc_info())
             raise
+        finally:
+            raw_server_sock.close()
+
+    def doit(ca, hostname, server_cert):
+        # socketpair and ssl don't work together on py2, because... reasons
+        #raw_client_sock, raw_server_sock = socket.socketpair()
+        listener = socket.socket()
+        listener.bind(("127.0.0.1", 0))
+        listener.listen(1)
+        raw_client_sock = socket.socket()
+        raw_client_sock.connect(listener.getsockname())
+        raw_server_sock, _ = listener.accept()
+        listener.close()
+        with ThreadPoolExecutor(2) as tpe:
+            f1 = tpe.submit(fake_ssl_client, ca, raw_client_sock, hostname)
+            f2 = tpe.submit(fake_ssl_server, server_cert, raw_server_sock)
+            f1.result()
```

Comment:
I think we're OK -- the only way this could lock up is something truly weird happens (like -- Python semantics violating weird), and there's not much that can be done to defend against that. Also I'm not sure how to do a timeout here (there's no way to cancel a thread). Worst case Travis times out after 10 minutes or whatever.